File: packages/ts-moose-lib/src/dmv2/sdk/cdcSource.ts
import { IJsonSchemaCollection } from "typia";
import { ClickHouseEngines } from "../../dataModels/types";
import { Column, DataType } from "../../dataModels/dataModelTypes";
import { OlapConfig, OlapTable } from "./olapTable";
import { Stream, StreamConfig } from "./stream";
import { LifeCycle } from "./lifeCycle";
import { TypedBase, TypiaValidators } from "../typedBase";
import { getMooseInternal } from "../internal";
import { getSourceLocationFromStack } from "../utils/stackTrace";

export type CdcOperation = "insert" | "update" | "delete";

export interface CdcEvent<T> {
  op: CdcOperation;
  before?: T | null;
  after?: T | null;
  ts: Date;
  lsn: string;
  source: string;
}

export type CdcRow<T> = T & {
  __cdc_op: CdcOperation;
  __cdc_lsn: string;
  __cdc_ts: Date;
  __cdc_is_deleted: boolean;
};

export type CdcSourceConfig = {
  kind: string;
  connection: string;
  metadata?: { description?: string };
  lifeCycle?: LifeCycle;
};

export type CdcTableConfig<T> = {
  sourceTable: string;
  primaryKey: Array<keyof T & string>;
  stream?: boolean | Omit<StreamConfig<CdcEvent<T>>, "destination">;
  table?: boolean | OlapConfig<CdcRow<T>>;
  snapshot?: "initial" | "never";
  version?: string;
  metadata?: { description?: string };
  lifeCycle?: LifeCycle;
};

export class CdcSource {
  readonly name: string;
  readonly config: CdcSourceConfig;
  readonly tables: Map<string, CdcTable<any>> = new Map();
  metadata?: {
    description?: string;
    source?: { file?: string; line?: string };
  };

  constructor(name: string, config: CdcSourceConfig) {
    this.name = name;
    this.config = config;

    this.metadata = config.metadata ? { ...config.metadata } : {};
    if (!this.metadata.source) {
      const stack = new Error().stack;
      const info = getSourceLocationFromStack(stack);
      if (info) {
        this.metadata.source = { file: info.file, line: info.line.toString() };
      }
    }

    getMooseInternal().cdcSources.set(name, this);
  }

  registerTable(table: CdcTable<any>) {
    this.tables.set(table.name, table);
  }
}

export class CdcTable<T> extends TypedBase<T, CdcTableConfig<T>> {
  readonly source: CdcSource;
  readonly sourceTable: string;
  stream?: Stream<CdcEvent<T>>;
  changes?: Stream<CdcEvent<T>>;
  table?: OlapTable<CdcRow<T>>;

  constructor(name: string, source: CdcSource, config: CdcTableConfig<T>);

  constructor(
    name: string,
    source: CdcSource,
    config: CdcTableConfig<T>,
    schema?: IJsonSchemaCollection.IV3_1,
    columns?: Column[],
    validators?: TypiaValidators<T>,
    allowExtraFields?: boolean,
  );

  constructor(
    name: string,
    source: CdcSource,
    config: CdcTableConfig<T>,
    schema?: IJsonSchemaCollection.IV3_1,
    columns?: Column[],
    validators?: TypiaValidators<T>,
    allowExtraFields?: boolean,
  ) {
    super(name, config, schema, columns, validators, allowExtraFields);

    this.source = source;
    this.sourceTable = config.sourceTable;
    source.registerTable(this);

    if (!config.primaryKey || config.primaryKey.length === 0) {
      throw new Error(
        `CdcTable '${name}' requires a non-empty primaryKey array.`,
      );
    }

    const metaColumnNames = new Set([
      "__cdc_op",
      "__cdc_lsn",
      "__cdc_ts",
      "__cdc_is_deleted",
    ]);
    for (const col of this.columnArray) {
      if (metaColumnNames.has(col.name)) {
        throw new Error(
          `CdcTable '${name}' uses reserved CDC metadata column name '${col.name}'.`,
        );
      }
    }

    if (config.stream !== false) {
      const eventSchema = buildCdcEventSchema(this.schema);
      const eventColumns = buildCdcEventColumns(this.columnArray);
      const streamConfig: StreamConfig<CdcEvent<T>> = {
        ...(typeof config.stream === "object" ? config.stream : {}),
        lifeCycle: config.lifeCycle,
        ...(config.version && { version: config.version }),
        metadata: config.metadata,
      };

      this.stream = new Stream(
        name,
        streamConfig,
        eventSchema,
        eventColumns,
        undefined,
        false,
      );
      this.changes = this.stream;
    }

    if (config.table) {
      const rowSchema = buildCdcRowSchema(this.schema);
      const rowColumns = buildCdcRowColumns(this.columnArray);
      const baseTableConfig = (
        typeof config.table === "object" ?
          {
            ...config.table,
            lifeCycle: config.table.lifeCycle ?? config.lifeCycle,
            ...(config.version && { version: config.version }),
          }
        : {
            engine: ClickHouseEngines.ReplacingMergeTree,
            orderByFields: config.primaryKey as Array<keyof CdcRow<T> & string>,
            ver: "__cdc_lsn",
            isDeleted: "__cdc_is_deleted",
            lifeCycle: config.lifeCycle,
            ...(config.version && { version: config.version }),
          }) as OlapConfig<CdcRow<T>>;

      if (
        "engine" in baseTableConfig &&
        baseTableConfig.engine === ClickHouseEngines.ReplacingMergeTree
      ) {
        const replacingConfig = baseTableConfig as {
          engine: ClickHouseEngines.ReplacingMergeTree;
          ver?: string;
          isDeleted?: string;
        };
        replacingConfig.ver = replacingConfig.ver ?? "__cdc_lsn";
        replacingConfig.isDeleted =
          replacingConfig.isDeleted ?? "__cdc_is_deleted";
      }

      if (
        "orderByFields" in baseTableConfig &&
        (!baseTableConfig.orderByFields ||
          baseTableConfig.orderByFields.length === 0)
      ) {
        baseTableConfig.orderByFields = config.primaryKey as Array<
          keyof CdcRow<T> & string
        >;
      }

      this.table = new OlapTable(
        name,
        baseTableConfig,
        rowSchema,
        rowColumns,
        undefined,
      );
    }
  }
}

const buildCdcEventSchema = (
  base: IJsonSchemaCollection.IV3_1,
): IJsonSchemaCollection.IV3_1 => {
  const baseSchema = base.schemas[0] as any;
  const nullableBase = {
    anyOf: [baseSchema, { type: "null" }],
  } as any;

  return {
    ...base,
    schemas: [
      {
        type: "object",
        properties: {
          op: { type: "string" },
          before: nullableBase,
          after: nullableBase,
          ts: { type: "string", format: "date-time" },
          lsn: { type: "string" },
          source: { type: "string" },
        },
        required: ["op", "ts", "lsn", "source"],
      } as any,
    ],
  } as IJsonSchemaCollection.IV3_1;
};

const buildCdcRowSchema = (
  base: IJsonSchemaCollection.IV3_1,
): IJsonSchemaCollection.IV3_1 => {
  const baseSchema = base.schemas[0] as any;
  return {
    ...base,
    schemas: [
      {
        allOf: [
          baseSchema,
          {
            type: "object",
            properties: {
              __cdc_op: { type: "string" },
              __cdc_lsn: { type: "string" },
              __cdc_ts: { type: "string", format: "date-time" },
              __cdc_is_deleted: { type: "boolean" },
            },
            required: ["__cdc_op", "__cdc_lsn", "__cdc_ts", "__cdc_is_deleted"],
          },
        ],
      } as any,
    ],
  } as IJsonSchemaCollection.IV3_1;
};

const buildCdcEventColumns = (columns: Column[]): Column[] => {
  const nestedBefore: DataType = {
    name: "before",
    columns,
    jwt: false,
  };
  const nestedAfter: DataType = {
    name: "after",
    columns,
    jwt: false,
  };

  return [
    {
      name: "op",
      data_type: "String",
      primary_key: false,
      required: true,
      unique: false,
      default: null,
      annotations: [],
      ttl: null,
      codec: null,
      materialized: null,
      comment: null,
    },
    {
      name: "before",
      data_type: { nullable: nestedBefore },
      primary_key: false,
      required: false,
      unique: false,
      default: null,
      annotations: [],
      ttl: null,
      codec: null,
      materialized: null,
      comment: null,
    },
    {
      name: "after",
      data_type: { nullable: nestedAfter },
      primary_key: false,
      required: false,
      unique: false,
      default: null,
      annotations: [],
      ttl: null,
      codec: null,
      materialized: null,
      comment: null,
    },
    {
      name: "ts",
      data_type: "DateTime",
      primary_key: false,
      required: true,
      unique: false,
      default: null,
      annotations: [],
      ttl: null,
      codec: null,
      materialized: null,
      comment: null,
    },
    {
      name: "lsn",
      data_type: "String",
      primary_key: false,
      required: true,
      unique: false,
      default: null,
      annotations: [],
      ttl: null,
      codec: null,
      materialized: null,
      comment: null,
    },
    {
      name: "source",
      data_type: "String",
      primary_key: false,
      required: true,
      unique: false,
      default: null,
      annotations: [],
      ttl: null,
      codec: null,
      materialized: null,
      comment: null,
    },
  ];
};

const buildCdcRowColumns = (columns: Column[]): Column[] => [
  ...columns,
  {
    name: "__cdc_op",
    data_type: "String",
    primary_key: false,
    required: true,
    unique: false,
    default: null,
    annotations: [],
    ttl: null,
    codec: null,
    materialized: null,
    comment: null,
  },
  {
    name: "__cdc_lsn",
    data_type: "String",
    primary_key: false,
    required: true,
    unique: false,
    default: null,
    annotations: [],
    ttl: null,
    codec: null,
    materialized: null,
    comment: null,
  },
  {
    name: "__cdc_ts",
    data_type: "DateTime",
    primary_key: false,
    required: true,
    unique: false,
    default: null,
    annotations: [],
    ttl: null,
    codec: null,
    materialized: null,
    comment: null,
  },
  {
    name: "__cdc_is_deleted",
    data_type: "UInt8",
    primary_key: false,
    required: true,
    unique: false,
    default: null,
    annotations: [],
    ttl: null,
    codec: null,
    materialized: null,
    comment: null,
  },
];

File: packages/py-moose-lib/moose_lib/dmv2/cdc_source.py
"""
CDC source definitions for Moose Data Model v2 (dmv2).

This module provides classes for defining CDC sources and CDC tables,
including typed CDC event streams and optional CDC destination tables.
"""

from __future__ import annotations

from datetime import datetime
from typing import Optional, Generic, Literal, TypeVar, Union, Any

from pydantic import BaseModel, create_model

from .types import TypedMooseResource, T
from .stream import Stream, StreamConfig
from .olap_table import OlapTable, OlapConfig
from .life_cycle import LifeCycle
from ._source_capture import get_source_file_from_stack
from ._registry import _cdc_sources
from ..blocks import ReplacingMergeTreeEngine

CdcOperation = Literal["insert", "update", "delete"]


class CdcEvent(BaseModel, Generic[T]):
    op: CdcOperation
    before: Optional[T] = None
    after: Optional[T] = None
    ts: datetime
    lsn: str
    source: str


class CdcRow(BaseModel, Generic[T]):
    __cdc_op: CdcOperation
    __cdc_lsn: str
    __cdc_ts: datetime
    __cdc_is_deleted: bool


class CdcSourceConfig(BaseModel):
    kind: str
    connection: str
    metadata: Optional[dict] = None
    life_cycle: Optional[LifeCycle] = None


class CdcTableConfig(BaseModel):
    source_table: str
    primary_key: list[str]
    stream: bool | StreamConfig = True
    table: bool | OlapConfig = True
    snapshot: Optional[Literal["initial", "never"]] = None
    version: Optional[str] = None
    metadata: Optional[dict] = None
    life_cycle: Optional[LifeCycle] = None


class CdcSource:
    def __init__(self, name: str, config: CdcSourceConfig):
        self.name = name
        self.config = config
        if config.metadata:
            self.metadata = (
                config.metadata.copy()
                if isinstance(config.metadata, dict)
                else config.metadata
            )
        else:
            self.metadata = {}

        if not isinstance(self.metadata, dict):
            self.metadata = {}

        if "source" not in self.metadata:
            source_file = get_source_file_from_stack()
            if source_file:
                self.metadata["source"] = {"file": source_file}
        self.tables: dict[str, CdcTable[Any]] = {}

        _cdc_sources[name] = self

    def register_table(self, table: "CdcTable[Any]"):
        self.tables[table.name] = table


class CdcTable(TypedMooseResource, Generic[T]):
    stream: Optional[Stream] = None
    changes: Optional[Stream] = None
    table: Optional[OlapTable] = None

    def __init__(self, name: str, source: CdcSource, config: CdcTableConfig, **kwargs):
        super().__init__()
        self._set_type(name, self._get_type(kwargs))

        if not config.primary_key:
            raise ValueError("CdcTable requires a non-empty primary_key list")

        self.name = name
        self.source = source
        self.config = config
        self.source_table = config.source_table
        if config.metadata:
            self.metadata = (
                config.metadata.copy()
                if isinstance(config.metadata, dict)
                else config.metadata
            )
        else:
            self.metadata = {}

        if not isinstance(self.metadata, dict):
            self.metadata = {}

        if "source" not in self.metadata:
            source_file = get_source_file_from_stack()
            if source_file:
                self.metadata["source"] = {"file": source_file}

        source.register_table(self)

        event_model = _create_cdc_event_model(self._t)
        row_model = _create_cdc_row_model(self._t)

        if config.stream is not False:
            stream_config = (
                config.stream
                if isinstance(config.stream, StreamConfig)
                else StreamConfig(life_cycle=config.life_cycle)
            )
            if config.version:
                stream_config.version = config.version
            if config.metadata:
                stream_config.metadata = config.metadata
            self.stream = Stream[event_model](name, stream_config, t=event_model)
            self.changes = self.stream

        if config.table:
            table_config = (
                config.table
                if isinstance(config.table, OlapConfig)
                else OlapConfig(life_cycle=config.life_cycle)
            )
            if table_config.engine is None:
                table_config.engine = ReplacingMergeTreeEngine(
                    ver="__cdc_lsn", is_deleted="__cdc_is_deleted"
                )
            if config.version:
                table_config.version = config.version
            if not table_config.order_by_fields:
                table_config.order_by_fields = config.primary_key
            self.table = OlapTable[row_model](name, table_config, t=row_model)


def _create_cdc_event_model(model: type[BaseModel]) -> type[BaseModel]:
    return create_model(
        f"{model.__name__}CdcEvent",
        op=(CdcOperation, ...),
        before=(Optional[model], None),
        after=(Optional[model], None),
        ts=(datetime, ...),
        lsn=(str, ...),
        source=(str, ...),
    )


def _create_cdc_row_model(model: type[BaseModel]) -> type[BaseModel]:
    return create_model(
        f"{model.__name__}CdcRow",
        __base__=model,
        __cdc_op=(CdcOperation, ...),
        __cdc_lsn=(str, ...),
        __cdc_ts=(datetime, ...),
        __cdc_is_deleted=(bool, ...),
    )
